engine/src/test/java/org/hibernate/validator/test/spi/nodenameprovider/AnnotationPropertyNodeNameProvider.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.test.spi.nodenameprovider;
+
+import java.io.Serializable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.util.Objects;
+
+import org.hibernate.validator.spi.nodenameprovider.JavaBeanProperty;
+import org.hibernate.validator.spi.nodenameprovider.Property;
+import org.hibernate.validator.spi.nodenameprovider.PropertyNodeNameProvider;
+
+/**
+ * An example of how a name can be resolved from an annotation
+ *
+ * @author Damir Alibegovic
+ */
+class AnnotationPropertyNodeNameProvider implements PropertyNodeNameProvider, Serializable {
+	private static final String VALUE = "value";
+
+	private final Class<? extends Annotation> annotationType;
+	private final String annotationMemberName;
+
+	private AnnotationPropertyNodeNameProvider(Class<? extends Annotation> annotationType, String annotationMemberName) {
+		Objects.requireNonNull( annotationType );
+
+		this.annotationType = annotationType;
+		this.annotationMemberName = annotationMemberName;
+	}
+
+	AnnotationPropertyNodeNameProvider(Class<? extends Annotation> annotationType) {
+		this( annotationType, VALUE );
+	}
+
+	@Override
+	public String getName(Property property) {
+		if ( property instanceof JavaBeanProperty ) {
+			return getJavaBeanPropertyName( (JavaBeanProperty) property );
+		}
+
+		return getDefaultName( property );
+	}
+
+	private String getJavaBeanPropertyName(JavaBeanProperty property) {
+		String resolvedName = property.getName();
+
+		Field field = getField( resolvedName, property.getDeclaringClass() );
+		if ( field != null && field.isAnnotationPresent( annotationType ) ) {
+			Annotation a = field.getAnnotation( annotationType );
+
+			try {
+				resolvedName = (String) a.annotationType().getMethod( annotationMemberName ).invoke( a );
```

Comment:
I know it's just an example but let's try to get it right:
- the method should be resolved in the constructor
- we should use a Logger with a proper error message, not `e.printStackTrace();`.

keep in mind that examples are copied a lot.

Reply:
Thanks for pointing that the examples are copied a lot, didn't know that, will refactor it.

Just one question, do we really want to do the method resolution eagerly, in the constructor? It's possible that this operation will not be needed, so it's maybe better to do it lazily?

Reply:
Hmmm. Yeah, it's really a trade-off. Reflection calls are usually costly so we try to avoid them as much as possible.

In this case, I think I would lean towards initializing it once and for all in the constructor as it could be called quite a lot of time. I mean if you have this sort of thing, the usage will probably be quite generalized.

But sure, it will depend of each use case so feel free to keep it that way.

Reply:
Ah, I see, your concern is performance related.

This class is used as a resolver of the property name (by using reflection), so each time a `getName(Property property)` is called, a different name is returned (because it is called with a different property parameter).

I just double checked, all of the Reflection calls that are done by the current implementation are actually necessary, so I think there is not much that can be improved.

So i would leave it as is.